src/main/java/com/khomishchak/cryproportfolio/services/integration/whitebit/WhiteBitServiceImpl.java

```diff
@@ -60,87 +65,115 @@ public class WhiteBitServiceImpl implements WhiteBitService {
     private final int retryMaxAttempts;
     private final Duration retryMinBackoff;
     private final MarketService marketService;
+    private final WhiteBitResponseMapper responseMapper;
 
     public WhiteBitServiceImpl(UserRepository userRepository, BalanceRepository balanceRepository,
             @Qualifier("WhiteBitApiWebClient") WebClient webClient, ApiKeySettingRepository apiKeySettingRepository,
             @Value("${crypto.portfolio.integration.exchanger.api.retry.maxAttempts:2}") int retryMaxAttempts,
             @Value("${crypto.portfolio.integration.exchanger.api.retry.minBackoffSeconds:2}") int retryMinBackoffSeconds,
-            MarketService marketService) {
+            MarketService marketService, WhiteBitResponseMapper responseMapper) {
         this.userRepository     = userRepository;
         this.apiKeySettingRepository = apiKeySettingRepository;
         this.balanceRepository  = balanceRepository;
         this.webClient          = webClient;
         this.retryMaxAttempts   = retryMaxAttempts;
         this.retryMinBackoff    = Duration.ofSeconds(retryMinBackoffSeconds);
         this.marketService      = marketService;
+        this.responseMapper     = responseMapper;
     }
 
     @Override
     public Balance getAccountBalance(long accoId) {
-        List<ApiKeySetting> apiKeys = apiKeySettingRepository.findAllByUserId(accoId);
-
-        ApiKeysPair keysPair = apiKeys.stream()
-                .filter(keys -> CODE.equals(keys.getCode()))
-                .map(ApiKeySetting::getApiKeys).findFirst().get();
-
+        ApiKeysPair keysPair = getApiKeysPair(accoId);
         String apiKey = keysPair.getPublicApi();
-
-        validateApiKey(apiKey, CODE);
+        validateApiKey(apiKey);
 
         String requestJson = String.format("{\"request\":\"%1$s\",\"nonce\":\"%2$s\",\"nonceWindow\":false}",
                 GET_MAIN_BALANCE_URL,
                 System.currentTimeMillis());
 
-        String payload = Base64.getEncoder().encodeToString(requestJson.getBytes());
-        String signature = null;
+        WhiteBitBalanceResp response =
+                makeWebPostRequest(GET_MAIN_BALANCE_URL, requestJson, keysPair, WhiteBitBalanceResp.class);
+        List<Currency> availableCurrencies = responseMapper.mapToCurrencies(response);
+
+        Balance balance = Balance.builder()
+                .code(CODE)
+                .user(userRepository.getReferenceById(accoId))
+                .totalValue(getTotalPrices(availableCurrencies))
+                .currencies(availableCurrencies)
+                .build();
+
+        return balanceRepository.save(balance);
+    }
+
+    @Override
+    public List<DepositWithdrawalTransaction> getDepositWithdrawalHistory(long accoId) {
+        ApiKeysPair keysPair = getApiKeysPair(accoId);
+        String apiKey = keysPair.getPublicApi();
+        validateApiKey(apiKey);
+
+        String requestJson = String.format(
```

Comment:
Why not to create an Java class representing the request/response and then convert to string using json mapper